service/operator/service.go

```diff
@@ -151,128 +147,353 @@ func (s *Service) Boot() {
+	// everything we find here is supposed to be created. Note that the creation
+	// of config-map data and service ports is always only an update operation
+	// against the Kubernetes API. Anyway, this concept here implements how a real
+	// reconciliation should drive specific parts of the current state towards the
+	// desired state, because a decent reconciliation is not always only an update
+	// operation of existing resources, but e.g. creation of resources. In our
+	// case here we only transform data within resources. Therefore the update.
+	{
+		// Process config-map to find its create state.
+		{
+			createState.ConfigMap.Data = map[string]string{}
+			for k, v := range dState.ConfigMapData {
+				if !inConfigMapData(createState.ConfigMap.Data, k, v) {
+					createState.ConfigMap.Data[k] = v
+				}
+			}
+		}
 
-	_, err = s.k8sClient.CoreV1().Services(s.namespace).Update(k8sService)
-	if err != nil {
-		return microerror.MaskAny(err)
+		// Process service to find its create state.
+		{
+			for _, p := range dState.ServicePorts {
+				if !inServicePorts(createState.Service.Spec.Ports, p) {
+					createState.Service.Spec.Ports = append(createState.Service.Spec.Ports, p)
+				}
+			}
+		}
 	}
 
-	return nil
+	return createState, nil
 }
 
-func (s *Service) deleteFunc(obj interface{}) {
-	s.mutex.Lock()
-	defer s.mutex.Unlock()
+func (s *Service) GetDeleteState(obj, currentState, desiredState interface{}) (interface{}, error) {
+	customObject, ok := obj.(*ingresstpr.CustomObject)
+	if !ok {
+		return nil, microerror.MaskAnyf(wrongTypeError, "expected '%T', got '%T'", &ingresstpr.CustomObject{}, obj)
+	}
+	cState, ok := currentState.(CurrentState)
+	if !ok {
+		return nil, microerror.MaskAnyf(wrongTypeError, "expected '%T', got '%T'", CurrentState{}, currentState)
+	}
+	dState, ok := desiredState.(DesiredState)
+	if !ok {
+		return nil, microerror.MaskAnyf(wrongTypeError, "expected '%T', got '%T'", DesiredState{}, desiredState)
+	}
 
-	err := s.deleteFuncError(obj)
-	if err != nil {
-		s.logger.Log("error", fmt.Sprintf("%#v", err), "event", "delete")
+	s.logger.Log("debug", "get delete state", "cluster", customObject.Spec.GuestCluster.ID)
+
+	var deleteState ActionState
+
+	// Make sure the current state of the Kubernetes resources is known by the
+	// delete action. The resources we already fetched represent the source of
+	// truth. They have to be used as base to actually update the resources in the
+	// next steps.
+	{
+		deleteState.ConfigMap = cState.ConfigMap
+		deleteState.Service = cState.Service
 	}
-}
 
-func (s *Service) deleteFuncError(obj interface{}) error {
-	return nil
+	// Find anything which is in current state but not in the desired state. This
+	// lets us drive the current state towards the desired state, because
+	// everything we find here is supposed to be deleted. Note that the deletion
+	// of config-map data and service ports is always only an update operation
```

Comment:
Exactly. Describing such complex functionalities in the code documentation is quite difficult.